### PR TITLE
test harness

### PR DESCRIPTION
**Pull request by jumanjiman**
this is not ready. i'm opening the PR to see how it fares in travis.

### CODE REVIEW

**Comment by purpleidea** [comment]
zomg, awesome :)

File: misc/bootstrap.sh [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	smitty git config --add remote.upstream.fetch '+refs/pull/*/head:refs/remotes/upstream/pr/*'
+fi
+
+# Install a specific version of gometalinter.
+go get -u gopkg.in/alecthomas/gometalinter.v1
+dir="$(dirname $(which gometalinter.v1))"
+mv "${dir}/gometalinter.v1" "${dir}/gometalinter"
+
+# Install or update go dependencies.
+smitty go get golang.org/x/tools/cmd/goimports
```
Comment (purpleidea) [inline misc/bootstrap.sh]: what's a smitty btw?

Comment (jumanjiman) [reply misc/bootstrap.sh]: it's a function that shows a command and its arguments, then runs the command.

i like to use it for wrapper scripts to passively educate users about what's happening behind the scenes. many users ignore it, but others see the commands and get curious.

example:

```
$ misc/bootstrap.sh 
[INFO] Running misc/bootstrap.sh
[INFO] git submodule sync
Synchronizing submodule url for 'vendor/github.com/coreos/etcd'
Synchronizing submodule url for 'vendor/github.com/grpc-ecosystem/go-grpc-prometheus'
Synchronizing submodule url for 'vendor/github.com/grpc-ecosystem/grpc-gateway'
Synchronizing submodule url for 'vendor/github.com/ugorji/go'
Synchronizing submodule url for 'vendor/google.golang.org/grpc'
Synchronizing submodule url for 'vendor/gopkg.in/fsnotify.v1'
Synchronizing submodule url for 'vendor/honnef.co/go/augeas'
[INFO] git submodule update --init --recursive
[WARN] About to clean git submodules.
Press ^C within 3s to abort.
[INFO] git clean -ffd
[INFO] upstream git remote is OK
[INFO] config to fetch upstream PR is OK
[INFO] go get golang.org/x/tools/cmd/goimports
[INFO] go get github.com/alecthomas/gometalinter
-snip-
```

It's an homage to AIX's [SMIT](https://www.ibm.com/developerworks/aix/library/au-smit/index.html), which has a gui for a new admin to click around in, then it shows the actual command and args. AIX has a tui version called [smitty](https://www.ibm.com/support/knowledgecenter/en/ssw_aix_71/com.ibm.aix.cmds5/smitty.htm).

I've always felt the AIX technique to allow a new admin to use a gui, but then show the actual CLI version, to be a useful teaching technique.

:-)

**Comment by purpleidea** [comment]
@jumanjiman How goes it?

**Comment by jumanjiman** [comment]
> How goes it?

everything is pretty straightforward outside of `test/shell`.
I'm currently trying to split the test script logic from the tests themselves to provide separation.

my vision for this experiment was to make the test harness  reproducible with minimal change to existing code.

a couple things i've noticed:

* acceptance tests need sudo access
* test harness assume /tmp is mounted exec

It would be easier if I could simply test each `test/shell/*` thing (acceptance test) in a throw-away docker container. 

Currently: 

```
if <running_in_travis>; then
  run acceptance tests, with side-effect to pollute local test environment (and make best effort to clean between tests)
else
  exit 0
fi
```

ideally:

```
if <docker_is_available>; then
  run each acceptance test in an ephemeral container and never pollute local environment
else
  warn "skipping mgmt acceptance tests since docker is unvailable"
fi
```

**Comment by purpleidea** [comment]
@jumanjiman Well I've had a few casual looks at this huge PR. it looks awesome. As for major changes like docker vs. not docker, it would be great to merge the obvious, less discuss-able things and get them in early so that it's easier to review, and the harder stuff doesn't block the easy wins :)

Thanks again!

**Comment by jumanjiman** [comment]
> As for major changes like docker vs. not docker, it would be great to merge the obvious, less discuss-able things and get them in early so that it's easier to review

yes. 

this PR is bigger than I had hoped, but at least it will be bisect-clean by the time i'm done, assuming it works.

**Comment by jumanjiman** [comment]
@purpleidea This is finally ready to discuss

**Comment by purpleidea** [comment]
@jumanjiman Awesome!!

I shall review shortly! Is it okay if I cherry-pick and merge the easy LGTM commits to make review of the rest easier?

**Comment by jumanjiman** [comment]
whatever is easier for you.
LMK if you'd like me to break the `test/util.sh` functions into a separate commit to make the cherry-picks clean.

**Comment by purpleidea** [comment]
Closing due to inactivity. Please send a new PR that applies to current master if there are remaining issues we can improve! Thanks!

**Pull request closed without merging**